### ISSUE TITLE
[BUG] Fleet agent high CPU usage after upgrade

### ISSUE THREAD

**Issue by janeczku** [harvester/harvester#2245]
**Describe the bug**
While upgrading from 1.0.0 to 1.0.1 the `cattle-fleet-agent` in the `cattle-fleet-local-system` namespace is consuming a lot of CPU cycles. It appears to be stuck in a loop going by the logs.

```
mike:/home/rancher # kubectl top pod -A
W0509 15:55:30.179544   32731 top_pod.go:140] Using json format to get metrics. Next release will switch to protocol-buffers, switch early by passing --use-protocol-buffers flag
NAMESPACE                   NAME                                                     CPU(cores)   MEMORY(bytes)
cattle-fleet-local-system   fleet-agent-57497ff7dc-596gs                             1612m        431Mi
cattle-fleet-system         fleet-agent-96f6f455c-74gnq                              2m           162Mi
cattle-fleet-system         fleet-controller-5746685958-wh4fv                        44m          186Mi
cattle-fleet-system         gitjob-cc9948fd7-87n7b                                   2m           126Mi
cattle-monitoring-system    prometheus-rancher-monitoring-prometheus-0               4m           80Mi
cattle-monitoring-system    rancher-monitoring-grafana-d9c56d79b-jnbpr               5m           207Mi
cattle-monitoring-system    rancher-monitoring-kube-state-metrics-5bc8bb48bd-psjq2   2m           73Mi
cattle-monitoring-system    rancher-monitoring-operator-559767d69b-zn59x             7m           302Mi
cattle-monitoring-system    rancher-monitoring-prometheus-adapter-8846d4757-hzvws    70m          294Mi
```

```time="2022-05-09T15:53:34Z" level=info msg="getting history for release harvester"
W0509 15:53:34.855326       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0509 15:53:34.989666       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
time="2022-05-09T15:53:35Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:35Z" level=info msg="getting history for release rancher-monitoring"
time="2022-05-09T15:53:35Z" level=info msg="getting history for release harvester"
W0509 15:53:36.431255       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
time="2022-05-09T15:53:36Z" level=error msg="bundle mcc-harvester: kubevirt.kubevirt.io harvester-system/kubevirt [progressing] Deploying version 0.49.0-2 with registry registry.suse.com/harvester-beta, Resource is current"
time="2022-05-09T15:53:36Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:37Z" level=info msg="getting history for release rancher-monitoring"
W0509 15:53:37.376645       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
time="2022-05-09T15:53:37Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:38Z" level=info msg="getting history for release harvester"
W0509 15:53:38.823240       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0509 15:53:38.841396       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
time="2022-05-09T15:53:39Z" level=error msg="bundle mcc-harvester: kubevirt.kubevirt.io harvester-system/kubevirt [progressing] Deploying version 0.49.0-2 with registry registry.suse.com/harvester-beta, Resource is current"
time="2022-05-09T15:53:39Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:39Z" level=error msg="bundle mcc-rancher-monitoring: daemonset.apps cattle-monitoring-system/rancher-monitoring-prometheus-node-exporter [progressing] Available: 3/4"
time="2022-05-09T15:53:39Z" level=info msg="getting history for release rancher-monitoring"
W0509 15:53:39.773264       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
time="2022-05-09T15:53:40Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:40Z" level=info msg="getting history for release harvester"
W0509 15:53:41.124003       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0509 15:53:41.137849       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
time="2022-05-09T15:53:41Z" level=error msg="bundle mcc-harvester: kubevirt.kubevirt.io harvester-system/kubevirt [progressing] Deploying version 0.49.0-2 with registry registry.suse.com/harvester-beta, Resource is current"
time="2022-05-09T15:53:41Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:41Z" level=info msg="getting history for release local-managed-system-agent"
time="2022-05-09T15:53:41Z" level=info msg="getting history for release rancher-monitoring"
time="2022-05-09T15:53:42Z" level=info msg="getting history for release local-managed-system-agent"
W0509 15:53:42.073228       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
time="2022-05-09T15:53:42Z" level=info msg="getting history for release local-managed-system-agent"
time="2022-05-09T15:53:42Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:42Z" level=info msg="getting history for release harvester"
time="2022-05-09T15:53:43Z" level=info msg="getting history for release rancher-monitoring"
W0509 15:53:43.495638       1 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
```

**Comment by thowd22** [comment]
Im seeing the same issue. I cant even start a VM because my cpu is running 3000-7000 percent over provisioned. This is a brand new install.

**Comment by w13915984028** [comment]
connect with 
#1893 
https://github.com/rancher/fleet/issues/760

Those issues are highly possible to be caused by same root cause.

**Comment by w13915984028** [comment]
The root cause analysis:
https://github.com/w13915984028/harvester-develop-summary/blob/main/fleet-agent-high-cpu-usage.md

**Comment by dweomer** [comment]
> The root cause analysis: https://github.com/w13915984028/harvester-develop-summary/blob/main/fleet-agent-high-cpu-usage.md

Based on the work done by @w13915984028 I should think you have two straight-forward options to consider by fixing fleet:
- ship, via serialization, full-fidelity status objects, down to every leaf node in the (oddly deep) object graph (aka no private struct fields)
- marhsal then unmarshal new/old objects prior to comparison via `equality.Semantic.DeepEqual` (wrangler used to do a lot of this kind of hackery)

**Comment by w13915984028** [comment]
@dweomer Hope you are doing well :)

Not sure you noticed, the `equality.Semantic.DeepEqual` is called from the auto-generated code in Rancher framework.
```
func (a *bundleDeploymentStatusHandler) sync
(key string, obj *v1alpha1.BundleDeployment) (*v1alpha1.BundleDeployment, error) {
```

When the framework can not fix it, then it needs to be solved in each caller. And most of us are not noticing this potential bug.

**Comment by dweomer** [comment]
It doesn't surprise me that the naive comparison is happening in generated code. Maybe fix the template?

It does surprise me that there are struct(s) in the graph with exported-but-not-serialized leaves/nodes _and_ that this has gone unnoticed until now.

**Comment by w13915984028** [comment]
connect: https://github.com/rancher/rancher/issues/38035

@dweomer 
Maybe Rancher framework need a customized `equality.Semantic.DeepEqual`, which excludes the private fields, to be used in the auto-generated code. Thus, the hack part is fully transparent to users.  I would prefer the framkework solve it.

I noticed for a long time, that even a blank Harvester single node cluster still consumes a noticable amount of CPU resources. Then focused on the issue and spend tens of hours to reproduce and dig. Finally found this BUG.

From the performance perspective, there could be more to optimize.

**Comment by dweomer** [comment]
> Maybe Rancher framework need a customized `equality.Semantic.DeepEqual`, which excludes the private fields, to be used in the auto-generated code. Thus, the hack part is fully transparent to users. I would prefer the framkework solve it.

Yeah, we are on the same page with this.

> I noticed for a long time, that even a blank Harvester single node cluster still consumes a noticable amount of CPU resources. Then focused on the issue and spend tens of hours to reproduce and dig. Finally found this BUG.

@w13915984028 I didn't mean to insinuate that you or anyone on the Harvester team was lax in any way whatsoever. You did excellent work to find the bug! Your write-up had all the context I or anyone familiar with fleet/wrangler code would need to understand the problem. (Assuming that my assessment, from my admittedly removed perch, is correct).

> From the performance perspective, there could be more to optimize.

Sure, but such should probably be considered out of scope.

**Comment by w13915984028** [comment]
@dweomer Great, we are aligned.

Regarding the topic of performance,  I did not express it clearly.

First, I noticed that even a newly installed Harvester cluster, without any user workloads, still takes notable amout of CPU resources. The fleet-agent issue is the right break point to dive into.

Next step, will continue investigate other parts as following, which are taken from a 2-node cluster without user workload, those PODs look busy.  As said before: `From the performance perspective, there could be more to optimize` , it is related to the whole system.

```
$ kubectl top pod -A 

NAMESPACE                   NAME                                                     CPU(cores)   MEMORY(bytes)   
          
cattle-monitoring-system    prometheus-rancher-monitoring-prometheus-0               60m          746Mi           
 ...
cattle-monitoring-system    rancher-monitoring-prometheus-adapter-8846d4757-srhm9    176m         135Mi           
...
cattle-system               rancher-6d977d6779-pz57x                                 9m           786Mi           
cattle-system               rancher-6d977d6779-xs2gf                                 8m           569Mi    
          
kube-system                 etcd-harv41                                              38m          558Mi           
kube-system                 kube-apiserver-harv41                                    172m         1709Mi          
kube-system                 kube-controller-manager-harv41                           12m          182Mi           

kube-system                 rke2-canal-5kt8v                                         13m          205Mi           
kube-system                 rke2-canal-jn66f                                         13m          201Mi           

kube-system                 rke2-coredns-rke2-coredns-autoscaler-7566b44b85-4vttb    18m          14Mi            

longhorn-system             longhorn-loop-device-cleaner-lx27m                       1m           2Mi             // normal , for comparing


longhorn-system             longhorn-manager-dp9r2                                   13m          91Mi            
longhorn-system             longhorn-manager-hbhkq                                   13m          88Mi            

longhorn-system             longhorn-ui-74d4bd689d-kkgtx                             0m           3Mi            // normal , for comparing
```

**Comment by w13915984028** [comment]
The root cause is same as #1893 , will be solved after bump fleet-agent to v0.5.0 (https://github.com/harvester/harvester/issues/3150)

close
